docs/proposals/proposal-001-trigger-and-deploy.md

```diff
@@ -0,0 +1,233 @@
+release and update the variable with the new version. This is to ensure each
+release is only triggered once.
+
+If sub components are specified then the pipeline will be triggered once per
+sub component.
+
+### Trigger
+
+The green reviews pipeline will be triggered by sending a [workflow_dispatch](https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event)
+event via the GitHub REST API.
+
+Inputs are
+
+- `cncf_project`: **required** Project to be deployed e.g. `falco`
+- `cncf_project_sub`: **optional** Subcomponent if project has multiple variants
+they wish to test e.g. Falco wish to test 3 falco drivers `modern-ebpf`, `kmod`
+and `ebpf`
+- `version`: **required** Version of project to be tested e.g. `0.37.0`
+
+```sh
+curl -X POST \
+     -H "Accept: application/vnd.github.v3+json" \
+     -H "Authorization: token $GITHUB_PAT" \
+     https://api.github.com/repos/cncf-tags/green-reviews-tooling/actions/workflows/pipeline.yaml/dispatches \
+     -d '{"ref":"0.2.0", "inputs": {"cncf_project": "falco", "cncf_project_sub": "modern-ebpf","version":"0.37.0"}}'
+```
+
+The pipeline is versioned by creating releases of the `green-reviews-tooling`
+repo. The git tag to use is passed via the `ref` param.
+
+The CNCF projects will be given a GitHub fine grained access token limited to
+the `green-reviews-tooling` repo. This token will have
+
+- Read access to code and metadata
+- Read write access to actions
+
+### Deploy
+
+Flux is used to deploy the CNCF project. Projects are able to use either
+`kustomization` or `helmrelease` resources to deploy their project.
+
+When the pipeline executes it will look for manifest files in the projects dir.
+If there is a manifest matching the `cncf_project` input its contents will be
+applied using kubectl. The same applies for the `cncf_project_sub` input. 
+
+The `version` param is injected into the files to ensure the correct version of
+the project is deployed. (For these minor changes we can utilize kustomize)
+
+```
+projects
+└── falco
+    ├── ebpf.yaml
+    ├── falco.yaml
+    ├── kmod.yaml
+    └── modern-ebpf.yaml
+```
+
+The pipeline will use a GitHub secret that has a kubeconfig to access the
+green reviews cluster. The deploy step in the pipeline will wait for the newly
+created flux resources to be reconciled before proceeding to the run step.
+
+We will have a node to deploy Falco and another to run the benchmarks
+so we will use [concurrency](https://docs.github.com/en/actions/using-jobs/using-concurrency)
+to only allow a single execution of the pipeline at any one time.
+
+The separate nodes are a best practice to prevent other components affecting
+the energy measurements.
+
+### Cleanup
+
+On completion of the pipeline run whether it was successful or failed the flux
+resources will be deleted via kubectl. The pipeline will wait for the flux
+resources to be deleted before exiting.
```

Comment:
This depends on which CLI was used to create the resources (see above for pros/cons of flux vs kubectl). If the resources were created using the `flux` CLI, it would be best to delete them using the `flux` CLI as well rather than `kubectl` since those resources would be controlled by Flux resources.

Reply:
> How do we trigger the delete? How about adding an action_type input to the workflow with deploy or uninstall?

> In the workflow we can have multiple jobs that run depending on the input. In the uninstall job we can add the flux delete command.

@rossf7 was not able to understand this part
even if we have running benchmark test it will trigger deletion once the benchmarking is completed. so may be we can start with this or something which you are refering to?

Reply:
> Let's add a note in the proposal that we should run a step that deletes the project that was deployed.

@nikimanoledaki I updated the Cleanup section to have more detail on the deletion. PTAL

> was not able to understand this part

@dipankardas011 Sorry for causing confusion. This was if we were going to delete with the flux CLI. However we decided to use kubectl so this doesn't apply.

Reply:
I was just thinking... how does the Trigger and Deploy pipeline know when the benchmarks are finished, in order to clean up? In general, how is the communication between the two?

Something related: when deploying the CNCF project (I suppose with `kubectl apply`) we should follow with a `kubectl wait --for condition`, so that the benchmarking only starts after everything is up and running, I suppose?

And for the benchmarks, per the last meeting discussion (maybe I should take this to the other PR):
- If the benchmark tests have a start and an end, there should be some kind of signal that this pipeline knows when to clean up (or perhaps it just knows to wait until the Benchark pipeline(s) are done?)
- If it's a test which runs on an interval of time, like the `stress-ng` for 15 minutes, who determines when to start and stop it? I suppose that is part of the Benchmarking step/pipeline to decide? But this is related to the unit R of the formula.

Maybe I'm complicating things and all of this is already clear, I'd like to hear everybody's thoughts..

Reply:
> how does the Trigger and Deploy pipeline know when the benchmarks are finished, in order to clean up? In general, how is the communication between the two?

@locomundo For a successful pipeline run this will be after the benchmarks have run and the metrics are saved to long term storage. I think this should be covered in proposal 3 but I added a note on that to the Cleanup section.

> Something related: when deploying the CNCF project (I suppose with kubectl apply) we should follow with a kubectl wait --for condition, so that the benchmarking only starts after everything is up and running, I suppose?

Yes we will use `kubectl apply -f` to apply the manifests but good point we'll also need to use `kubectl wait` before starting the benchmarking to ensure flux has reconciled the resources. I added this to the Deploy section.

> If it's a test which runs on an interval of time, like the stress-ng for 15 minutes, who determines when to start and stop it? I suppose that is part of the Benchmarking step/pipeline to decide? But this is related to the unit R of the formula.

I think the benchmarking step will need to control this but I think its best to include that in the proposal 2 PR.